{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# DataPath Example 4\n",
    "\n",
    "This notebook covers somewhat more advanced examples for using `DataPath`s. It assumes that you understand \n",
    "the concepts presented in the previous example notebooks.\n",
    "\n",
    "You should also read the ERMrest documentation and the derivapy wiki. There are more advanced concepts in this notebook that are demonstrated but not fully (re)explained here, as the concepts are explained in other documentation.\n",
    "\n",
    "## Exampe Data Model\n",
    "The examples require that you understand a little bit about the example catalog data model, which in this case manages data for biological experiments.\n",
    "\n",
    "### Key tables\n",
    "- `'dataset'` : represents a unit of data usually for a study or set of experiments;\n",
    "- `'biosample'` : a biosample (describes biological details of a specimen);\n",
    "- `'replicate'` : a replicate (describes both bio- and technical-replicates);\n",
    "- `'experiment'` : a bioassay (any type of experiment or assay; e.g., imaging, RNA-seq, ChIP-seq, etc.).\n",
    "\n",
    "### Relationships\n",
    "- `dataset <- biosample`: A dataset may have one to many biosamples. I.e., there is a \n",
    "  foreign key reference from biosample to dataset.\n",
    "- `dataset <- experiment`: A dataset may have one to many experiments. I.e., there \n",
    "  is a foreign key reference from experiment to dataset.\n",
    "- `experiment <- replicate`: An experiment may have one to many replicates. I.e., there is a\n",
    "  foreign key reference from replicate to experiment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import deriva modules and pandas DataFrame (for use in examples only)\n",
    "from deriva.core import ErmrestCatalog, get_credential\n",
    "from pandas import DataFrame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Connect with the deriva catalog\n",
    "protocol = 'https'\n",
    "hostname = 'www.facebase.org'\n",
    "catalog_number = 1\n",
    "credential = None\n",
    "# If you need to authenticate, use Deriva Auth agent and get the credential\n",
    "# credential = get_credential(hostname)\n",
    "catalog = ErmrestCatalog(protocol, hostname, catalog_number, credential)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the path builder interface for this catalog\n",
    "pb = catalog.getPathBuilder()\n",
    "\n",
    "# Get some local variable handles to tables for convenience\n",
    "dataset = pb.isa.dataset\n",
    "experiment = pb.isa.experiment\n",
    "biosample = pb.isa.biosample\n",
    "replicate = pb.isa.replicate"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Implicit DataPaths\n",
    "**Proceed with caution**\n",
    "\n",
    "For compactness, `Table` objects (and `TableAlias` objects) provide `DataPath`-like methods. E.g., `link(...)`, `filter(...)`, and `entities(...)`, which will implicitly create `DataPath`s rooted at the table and return the newly created path. These operations `return` the new `DataPath` rather than mutating the `Table` (or `TableAlias`) objects."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "855"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "entities = dataset.filter(dataset.released == True).entities()\n",
    "len(entities)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DataPath-like methods\n",
    "The `DataPath`-like methods on `Table`s are essentially \"wrapper\" functions over the implicitly generated `DataPath` rooted at the `Table` instance. The wrappers include, `link(...)`, `filter(...)`, `entities(...)`, `attributes(...)`, `aggregates(...)`, and `groupby(...)`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Attribute Examples"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Example: selecting all columns of a table instance\n",
    "Passing a table (or table instance) object to the `attributes(...)` method will project all (i.e., `*`) of its attributes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3272\n",
      "https://www.facebase.org/ermrest/catalog/1/attribute/D:=isa:dataset/experiment:=isa:experiment/replicate:=isa:replicate/D:*\n"
     ]
    }
   ],
   "source": [
    "path = dataset.alias('D').path\n",
    "path.link(experiment).link(replicate)\n",
    "results = path.attributes(path.D)\n",
    "print(len(results))\n",
    "print(results.uri)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It is important to remember that the `attributes(...)` method returns a result set based on the entity type of the last elmenent of the path. In this example that means the number of results will be determined by the number of unique rows in the replicate table instance in the path created above, as the last link method used the replicate table."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Example: selecting from multiple table instances\n",
    "More than one table instance may be selected in this manner and it can be mixed and matched with columns from other tables instances. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3272\n",
      "https://www.facebase.org/ermrest/catalog/1/attribute/D:=isa:dataset/experiment:=isa:experiment/replicate:=isa:replicate/D:*,experiment:experiment_type,replicate:*\n"
     ]
    }
   ],
   "source": [
    "results = path.attributes(path.D,\n",
    "                          path.experiment.experiment_type,\n",
    "                          path.replicate)\n",
    "print(len(results))\n",
    "print(results.uri)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " If you want to base the results on a different entity, you can introduce a table instance alias into the end of the path, before calling the attributes function.  In this case, even though we are asking for the same attributes, we are getting the set of datasets, not the set of replicates.  Also, since we are including the attributes from dataset in our query, we know that we will not be seeing any duplicate rows."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "142\n",
      "https://www.facebase.org/ermrest/catalog/1/attribute/D:=isa:dataset/experiment:=isa:experiment/replicate:=isa:replicate/$D/D:*,experiment:experiment_type,replicate:*\n"
     ]
    }
   ],
   "source": [
    "results = path.D.attributes(path.D,\n",
    "                            path.experiment.experiment_type,\n",
    "                            path.replicate)\n",
    "print(len(results))\n",
    "print(results.uri)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filtering Examples"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Example: filter on `null` attribute\n",
    "To test for a `null` attribute value, do an equality comparison against the `None` identity."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://www.facebase.org/ermrest/catalog/1/entity/dataset:=isa:dataset/experiment:=isa:experiment/molecule_type::null::\n",
      "438\n"
     ]
    }
   ],
   "source": [
    "path = dataset.link(experiment).filter(experiment.molecule_type == None)\n",
    "print(path.uri)\n",
    "print(len(path.entities()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Example: advanced text filters\n",
    "Deriva supports advanced text filters for regular expressions (`regexp`), case-instansitive regexp (`ciregexp`), and text search (`ts`). You may have to review your text and full-text indexes in your ERMrest catalog before using these features."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://www.facebase.org/ermrest/catalog/1/entity/dataset:=isa:dataset/description::ciregexp::palate\n",
      "40\n"
     ]
    }
   ],
   "source": [
    "path = dataset.filter(dataset.description.ciregexp('palate'))\n",
    "print(path.uri)\n",
    "print(len(path.entities()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Example: negate a filter\n",
    "Use the \"inverse\" ('`~`') operator to negate a filter. Negation works against simple comparison filters as demonstrated above as well as on logical operators to be discussed next. You must wrap the comparison or logical operators in an extra parens to use the negate operation, e.g., \"`~ (...)`\". "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://www.facebase.org/ermrest/catalog/1/entity/dataset:=isa:dataset/!(description::ciregexp::palate)\n",
      "811\n"
     ]
    }
   ],
   "source": [
    "path = dataset.filter( ~ (dataset.description.ciregexp('palate')) )\n",
    "print(path.uri)\n",
    "print(len(path.entities()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Example: filters with logical operators\n",
    "This example shows how to combine two comparisons with a conjuncting (i.e., `and` operator). Because Python's logical-and (`and`) keyword cannot be overloaded, we instead overload the bitwise-and (`&`) operator. This approach has become customary among many similar data access libraries."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://www.facebase.org/ermrest/catalog/1/entity/dataset:=isa:dataset/biosample:=isa:biosample/(species=NCBITAXON%3A10090)&(anatomy=UBERON%3A0002490)\n"
     ]
    }
   ],
   "source": [
    "path = dataset.link(biosample).filter(\n",
    "    ((biosample.species == 'NCBITAXON:10090') & (biosample.anatomy == 'UBERON:0002490')))\n",
    "\n",
    "print(path.uri)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RID</th>\n",
       "      <th>dataset</th>\n",
       "      <th>local_identifier</th>\n",
       "      <th>species</th>\n",
       "      <th>specimen</th>\n",
       "      <th>gene</th>\n",
       "      <th>genotype</th>\n",
       "      <th>strain</th>\n",
       "      <th>mutation</th>\n",
       "      <th>stage</th>\n",
       "      <th>...</th>\n",
       "      <th>gender</th>\n",
       "      <th>litter</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>_keywords</th>\n",
       "      <th>RCB</th>\n",
       "      <th>RMB</th>\n",
       "      <th>RCT</th>\n",
       "      <th>RMT</th>\n",
       "      <th>cell_source</th>\n",
       "      <th>cell_characterization</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-4TT8</td>\n",
       "      <td>1-4TSP</td>\n",
       "      <td>scWFE18_S197</td>\n",
       "      <td>NCBITAXON:10090</td>\n",
       "      <td>FACEBASE:1-4GNR</td>\n",
       "      <td>None</td>\n",
       "      <td>FACEBASE:1-4GV2</td>\n",
       "      <td>FACEBASE:1-4GYR</td>\n",
       "      <td>None</td>\n",
       "      <td>FACEBASE:1-4GJA</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>https://auth.globus.org/8ae274db-d033-47eb-bd3...</td>\n",
       "      <td>https://auth.globus.org/8ae274db-d033-47eb-bd3...</td>\n",
       "      <td>2018-12-10T19:42:13.870864-08:00</td>\n",
       "      <td>2018-12-10T19:43:19.779894-08:00</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2XEJ</td>\n",
       "      <td>TJP</td>\n",
       "      <td>A8IF4SM</td>\n",
       "      <td>NCBITAXON:10090</td>\n",
       "      <td>FACEBASE:1-4GNG</td>\n",
       "      <td>None</td>\n",
       "      <td>FACEBASE:1-4GT6</td>\n",
       "      <td>FACEBASE:1-4GYR</td>\n",
       "      <td>None</td>\n",
       "      <td>FACEBASE:1-4GJA</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>3/30/15AL1-5</td>\n",
       "      <td>2016-06-20</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>https://auth.globus.org/b506963e-d274-11e5-99f...</td>\n",
       "      <td>2018-03-12T18:12:27.599487-07:00</td>\n",
       "      <td>2018-11-27T16:53:36.379287-08:00</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2XEP</td>\n",
       "      <td>TJP</td>\n",
       "      <td>A8IF1SM</td>\n",
       "      <td>NCBITAXON:10090</td>\n",
       "      <td>FACEBASE:1-4GNG</td>\n",
       "      <td>None</td>\n",
       "      <td>FACEBASE:1-4GT6</td>\n",
       "      <td>FACEBASE:1-4GYR</td>\n",
       "      <td>None</td>\n",
       "      <td>FACEBASE:1-4GJA</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>9/29/14AL1-5</td>\n",
       "      <td>2016-06-20</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>https://auth.globus.org/b506963e-d274-11e5-99f...</td>\n",
       "      <td>2018-03-12T18:12:27.599487-07:00</td>\n",
       "      <td>2018-11-27T16:53:36.068132-08:00</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2XW6</td>\n",
       "      <td>TJP</td>\n",
       "      <td>A8IF4FR</td>\n",
       "      <td>NCBITAXON:10090</td>\n",
       "      <td>FACEBASE:1-4GNG</td>\n",
       "      <td>None</td>\n",
       "      <td>FACEBASE:1-4GT6</td>\n",
       "      <td>FACEBASE:1-4GYR</td>\n",
       "      <td>None</td>\n",
       "      <td>FACEBASE:1-4GJA</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>3/30/15AL1-5</td>\n",
       "      <td>2016-06-20</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>https://auth.globus.org/b506963e-d274-11e5-99f...</td>\n",
       "      <td>2018-03-12T18:12:27.599487-07:00</td>\n",
       "      <td>2018-11-27T16:53:36.644948-08:00</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2YAT</td>\n",
       "      <td>TJP</td>\n",
       "      <td>A8IF3FR</td>\n",
       "      <td>NCBITAXON:10090</td>\n",
       "      <td>FACEBASE:1-4GNG</td>\n",
       "      <td>None</td>\n",
       "      <td>FACEBASE:1-4GT6</td>\n",
       "      <td>FACEBASE:1-4GYR</td>\n",
       "      <td>None</td>\n",
       "      <td>FACEBASE:1-4GJA</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>3/30/15AL1-1</td>\n",
       "      <td>2016-06-20</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>https://auth.globus.org/b506963e-d274-11e5-99f...</td>\n",
       "      <td>2018-03-12T18:12:27.599487-07:00</td>\n",
       "      <td>2018-11-27T16:53:35.57911-08:00</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>3Q1E</td>\n",
       "      <td>VHE</td>\n",
       "      <td>W8IF1SM</td>\n",
       "      <td>NCBITAXON:10090</td>\n",
       "      <td>FACEBASE:1-4GNG</td>\n",
       "      <td>None</td>\n",
       "      <td>FACEBASE:1-4GV2</td>\n",
       "      <td>FACEBASE:1-4GYR</td>\n",
       "      <td>None</td>\n",
       "      <td>FACEBASE:1-4GJA</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>4/15/14WL1-1</td>\n",
       "      <td>2014-04-15</td>\n",
       "      <td>None</td>\n",
       "      <td>https://auth.globus.org/b506963e-d274-11e5-99f...</td>\n",
       "      <td>https://auth.globus.org/b506963e-d274-11e5-99f...</td>\n",
       "      <td>2018-03-16T12:03:38.228586-07:00</td>\n",
       "      <td>2018-11-27T16:53:35.681942-08:00</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>3Q1J</td>\n",
       "      <td>VHE</td>\n",
       "      <td>W8IF2SM</td>\n",
       "      <td>NCBITAXON:10090</td>\n",
       "      <td>FACEBASE:1-4GNG</td>\n",
       "      <td>None</td>\n",
       "      <td>FACEBASE:1-4GV2</td>\n",
       "      <td>FACEBASE:1-4GYR</td>\n",
       "      <td>None</td>\n",
       "      <td>FACEBASE:1-4GJA</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>4/15/14WL1-2</td>\n",
       "      <td>2014-04-15</td>\n",
       "      <td>None</td>\n",
       "      <td>https://auth.globus.org/b506963e-d274-11e5-99f...</td>\n",
       "      <td>https://auth.globus.org/b506963e-d274-11e5-99f...</td>\n",
       "      <td>2018-03-16T12:03:38.228586-07:00</td>\n",
       "      <td>2018-11-27T16:53:35.473682-08:00</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>3Q1P</td>\n",
       "      <td>VHE</td>\n",
       "      <td>W8IF3SM</td>\n",
       "      <td>NCBITAXON:10090</td>\n",
       "      <td>FACEBASE:1-4GNG</td>\n",
       "      <td>None</td>\n",
       "      <td>FACEBASE:1-4GV2</td>\n",
       "      <td>FACEBASE:1-4GYR</td>\n",
       "      <td>None</td>\n",
       "      <td>FACEBASE:1-4GJA</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>4/14/14WL1-6</td>\n",
       "      <td>2014-04-14</td>\n",
       "      <td>None</td>\n",
       "      <td>https://auth.globus.org/b506963e-d274-11e5-99f...</td>\n",
       "      <td>https://auth.globus.org/b506963e-d274-11e5-99f...</td>\n",
       "      <td>2018-03-16T12:03:38.228586-07:00</td>\n",
       "      <td>2018-11-27T16:53:35.473682-08:00</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>3Q1T</td>\n",
       "      <td>VHE</td>\n",
       "      <td>W8IF4SM</td>\n",
       "      <td>NCBITAXON:10090</td>\n",
       "      <td>FACEBASE:1-4GNG</td>\n",
       "      <td>None</td>\n",
       "      <td>FACEBASE:1-4GV2</td>\n",
       "      <td>FACEBASE:1-4GYR</td>\n",
       "      <td>None</td>\n",
       "      <td>FACEBASE:1-4GJA</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>4/15/14WL1-7</td>\n",
       "      <td>2014-04-15</td>\n",
       "      <td>None</td>\n",
       "      <td>https://auth.globus.org/b506963e-d274-11e5-99f...</td>\n",
       "      <td>https://auth.globus.org/b506963e-d274-11e5-99f...</td>\n",
       "      <td>2018-03-16T12:03:38.228586-07:00</td>\n",
       "      <td>2018-11-27T16:53:35.473682-08:00</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>3Q1Y</td>\n",
       "      <td>VHE</td>\n",
       "      <td>W8IF5SM</td>\n",
       "      <td>NCBITAXON:10090</td>\n",
       "      <td>FACEBASE:1-4GNG</td>\n",
       "      <td>None</td>\n",
       "      <td>FACEBASE:1-4GV2</td>\n",
       "      <td>FACEBASE:1-4GYR</td>\n",
       "      <td>None</td>\n",
       "      <td>FACEBASE:1-4GJA</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>4/28/14WL1-1</td>\n",
       "      <td>2014-04-28</td>\n",
       "      <td>None</td>\n",
       "      <td>https://auth.globus.org/b506963e-d274-11e5-99f...</td>\n",
       "      <td>https://auth.globus.org/b506963e-d274-11e5-99f...</td>\n",
       "      <td>2018-03-16T12:03:38.228586-07:00</td>\n",
       "      <td>2018-11-27T16:53:35.473682-08:00</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       RID dataset local_identifier          species         specimen  gene  \\\n",
       "0   1-4TT8  1-4TSP     scWFE18_S197  NCBITAXON:10090  FACEBASE:1-4GNR  None   \n",
       "1     2XEJ     TJP          A8IF4SM  NCBITAXON:10090  FACEBASE:1-4GNG  None   \n",
       "2     2XEP     TJP          A8IF1SM  NCBITAXON:10090  FACEBASE:1-4GNG  None   \n",
       "3     2XW6     TJP          A8IF4FR  NCBITAXON:10090  FACEBASE:1-4GNG  None   \n",
       "4     2YAT     TJP          A8IF3FR  NCBITAXON:10090  FACEBASE:1-4GNG  None   \n",
       "..     ...     ...              ...              ...              ...   ...   \n",
       "57    3Q1E     VHE          W8IF1SM  NCBITAXON:10090  FACEBASE:1-4GNG  None   \n",
       "58    3Q1J     VHE          W8IF2SM  NCBITAXON:10090  FACEBASE:1-4GNG  None   \n",
       "59    3Q1P     VHE          W8IF3SM  NCBITAXON:10090  FACEBASE:1-4GNG  None   \n",
       "60    3Q1T     VHE          W8IF4SM  NCBITAXON:10090  FACEBASE:1-4GNG  None   \n",
       "61    3Q1Y     VHE          W8IF5SM  NCBITAXON:10090  FACEBASE:1-4GNG  None   \n",
       "\n",
       "           genotype           strain mutation            stage  ... gender  \\\n",
       "0   FACEBASE:1-4GV2  FACEBASE:1-4GYR     None  FACEBASE:1-4GJA  ...   None   \n",
       "1   FACEBASE:1-4GT6  FACEBASE:1-4GYR     None  FACEBASE:1-4GJA  ...   None   \n",
       "2   FACEBASE:1-4GT6  FACEBASE:1-4GYR     None  FACEBASE:1-4GJA  ...   None   \n",
       "3   FACEBASE:1-4GT6  FACEBASE:1-4GYR     None  FACEBASE:1-4GJA  ...   None   \n",
       "4   FACEBASE:1-4GT6  FACEBASE:1-4GYR     None  FACEBASE:1-4GJA  ...   None   \n",
       "..              ...              ...      ...              ...  ...    ...   \n",
       "57  FACEBASE:1-4GV2  FACEBASE:1-4GYR     None  FACEBASE:1-4GJA  ...   None   \n",
       "58  FACEBASE:1-4GV2  FACEBASE:1-4GYR     None  FACEBASE:1-4GJA  ...   None   \n",
       "59  FACEBASE:1-4GV2  FACEBASE:1-4GYR     None  FACEBASE:1-4GJA  ...   None   \n",
       "60  FACEBASE:1-4GV2  FACEBASE:1-4GYR     None  FACEBASE:1-4GJA  ...   None   \n",
       "61  FACEBASE:1-4GV2  FACEBASE:1-4GYR     None  FACEBASE:1-4GJA  ...   None   \n",
       "\n",
       "          litter collection_date _keywords  \\\n",
       "0           None            None      None   \n",
       "1   3/30/15AL1-5      2016-06-20      None   \n",
       "2   9/29/14AL1-5      2016-06-20      None   \n",
       "3   3/30/15AL1-5      2016-06-20      None   \n",
       "4   3/30/15AL1-1      2016-06-20      None   \n",
       "..           ...             ...       ...   \n",
       "57  4/15/14WL1-1      2014-04-15      None   \n",
       "58  4/15/14WL1-2      2014-04-15      None   \n",
       "59  4/14/14WL1-6      2014-04-14      None   \n",
       "60  4/15/14WL1-7      2014-04-15      None   \n",
       "61  4/28/14WL1-1      2014-04-28      None   \n",
       "\n",
       "                                                  RCB  \\\n",
       "0   https://auth.globus.org/8ae274db-d033-47eb-bd3...   \n",
       "1                                                None   \n",
       "2                                                None   \n",
       "3                                                None   \n",
       "4                                                None   \n",
       "..                                                ...   \n",
       "57  https://auth.globus.org/b506963e-d274-11e5-99f...   \n",
       "58  https://auth.globus.org/b506963e-d274-11e5-99f...   \n",
       "59  https://auth.globus.org/b506963e-d274-11e5-99f...   \n",
       "60  https://auth.globus.org/b506963e-d274-11e5-99f...   \n",
       "61  https://auth.globus.org/b506963e-d274-11e5-99f...   \n",
       "\n",
       "                                                  RMB  \\\n",
       "0   https://auth.globus.org/8ae274db-d033-47eb-bd3...   \n",
       "1   https://auth.globus.org/b506963e-d274-11e5-99f...   \n",
       "2   https://auth.globus.org/b506963e-d274-11e5-99f...   \n",
       "3   https://auth.globus.org/b506963e-d274-11e5-99f...   \n",
       "4   https://auth.globus.org/b506963e-d274-11e5-99f...   \n",
       "..                                                ...   \n",
       "57  https://auth.globus.org/b506963e-d274-11e5-99f...   \n",
       "58  https://auth.globus.org/b506963e-d274-11e5-99f...   \n",
       "59  https://auth.globus.org/b506963e-d274-11e5-99f...   \n",
       "60  https://auth.globus.org/b506963e-d274-11e5-99f...   \n",
       "61  https://auth.globus.org/b506963e-d274-11e5-99f...   \n",
       "\n",
       "                                 RCT                               RMT  \\\n",
       "0   2018-12-10T19:42:13.870864-08:00  2018-12-10T19:43:19.779894-08:00   \n",
       "1   2018-03-12T18:12:27.599487-07:00  2018-11-27T16:53:36.379287-08:00   \n",
       "2   2018-03-12T18:12:27.599487-07:00  2018-11-27T16:53:36.068132-08:00   \n",
       "3   2018-03-12T18:12:27.599487-07:00  2018-11-27T16:53:36.644948-08:00   \n",
       "4   2018-03-12T18:12:27.599487-07:00   2018-11-27T16:53:35.57911-08:00   \n",
       "..                               ...                               ...   \n",
       "57  2018-03-16T12:03:38.228586-07:00  2018-11-27T16:53:35.681942-08:00   \n",
       "58  2018-03-16T12:03:38.228586-07:00  2018-11-27T16:53:35.473682-08:00   \n",
       "59  2018-03-16T12:03:38.228586-07:00  2018-11-27T16:53:35.473682-08:00   \n",
       "60  2018-03-16T12:03:38.228586-07:00  2018-11-27T16:53:35.473682-08:00   \n",
       "61  2018-03-16T12:03:38.228586-07:00  2018-11-27T16:53:35.473682-08:00   \n",
       "\n",
       "   cell_source cell_characterization  \n",
       "0         None                  None  \n",
       "1         None                  None  \n",
       "2         None                  None  \n",
       "3         None                  None  \n",
       "4         None                  None  \n",
       "..         ...                   ...  \n",
       "57        None                  None  \n",
       "58        None                  None  \n",
       "59        None                  None  \n",
       "60        None                  None  \n",
       "61        None                  None  \n",
       "\n",
       "[62 rows x 24 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DataFrame(path.entities())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Example: combine conjunction and disjunctions in filters\n",
    "Similar to the prior example, the filters allow combining of conjunctive and disjunctive operators. Like the bitwise-and operator, we also overload the bitwise-or (` | `) operator because the logical-or (`or`) operatar cannot be overloaded."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://www.facebase.org/ermrest/catalog/1/entity/dataset:=isa:dataset/biosample:=isa:biosample/((species=NCBITAXON%3A10090)&(anatomy=UBERON%3A0002490));((specimen=FACEBASE%3A1-4GNR)&(stage=FACEBASE%3A1-4GJA))\n"
     ]
    }
   ],
   "source": [
    "path = dataset.link(biosample).filter(\n",
    "    ((biosample.species == 'NCBITAXON:10090') & (biosample.anatomy == 'UBERON:0002490')) |\n",
    "    ((biosample.specimen == 'FACEBASE:1-4GNR') & (biosample.stage == 'FACEBASE:1-4GJA')))\n",
    "\n",
    "print(path.uri)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RID</th>\n",
       "      <th>dataset</th>\n",
       "      <th>local_identifier</th>\n",
       "      <th>species</th>\n",
       "      <th>specimen</th>\n",
       "      <th>gene</th>\n",
       "      <th>genotype</th>\n",
       "      <th>strain</th>\n",
       "      <th>mutation</th>\n",
       "      <th>stage</th>\n",
       "      <th>...</th>\n",
       "      <th>gender</th>\n",
       "      <th>litter</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>_keywords</th>\n",
       "      <th>RCB</th>\n",
       "      <th>RMB</th>\n",
       "      <th>RCT</th>\n",
       "      <th>RMT</th>\n",
       "      <th>cell_source</th>\n",
       "      <th>cell_characterization</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-4TT8</td>\n",
       "      <td>1-4TSP</td>\n",
       "      <td>scWFE18_S197</td>\n",
       "      <td>NCBITAXON:10090</td>\n",
       "      <td>FACEBASE:1-4GNR</td>\n",
       "      <td>None</td>\n",
       "      <td>FACEBASE:1-4GV2</td>\n",
       "      <td>FACEBASE:1-4GYR</td>\n",
       "      <td>None</td>\n",
       "      <td>FACEBASE:1-4GJA</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>https://auth.globus.org/8ae274db-d033-47eb-bd3...</td>\n",
       "      <td>https://auth.globus.org/8ae274db-d033-47eb-bd3...</td>\n",
       "      <td>2018-12-10T19:42:13.870864-08:00</td>\n",
       "      <td>2018-12-10T19:43:19.779894-08:00</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2XEJ</td>\n",
       "      <td>TJP</td>\n",
       "      <td>A8IF4SM</td>\n",
       "      <td>NCBITAXON:10090</td>\n",
       "      <td>FACEBASE:1-4GNG</td>\n",
       "      <td>None</td>\n",
       "      <td>FACEBASE:1-4GT6</td>\n",
       "      <td>FACEBASE:1-4GYR</td>\n",
       "      <td>None</td>\n",
       "      <td>FACEBASE:1-4GJA</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>3/30/15AL1-5</td>\n",
       "      <td>2016-06-20</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>https://auth.globus.org/b506963e-d274-11e5-99f...</td>\n",
       "      <td>2018-03-12T18:12:27.599487-07:00</td>\n",
       "      <td>2018-11-27T16:53:36.379287-08:00</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2XEP</td>\n",
       "      <td>TJP</td>\n",
       "      <td>A8IF1SM</td>\n",
       "      <td>NCBITAXON:10090</td>\n",
       "      <td>FACEBASE:1-4GNG</td>\n",
       "      <td>None</td>\n",
       "      <td>FACEBASE:1-4GT6</td>\n",
       "      <td>FACEBASE:1-4GYR</td>\n",
       "      <td>None</td>\n",
       "      <td>FACEBASE:1-4GJA</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>9/29/14AL1-5</td>\n",
       "      <td>2016-06-20</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>https://auth.globus.org/b506963e-d274-11e5-99f...</td>\n",
       "      <td>2018-03-12T18:12:27.599487-07:00</td>\n",
       "      <td>2018-11-27T16:53:36.068132-08:00</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2XW6</td>\n",
       "      <td>TJP</td>\n",
       "      <td>A8IF4FR</td>\n",
       "      <td>NCBITAXON:10090</td>\n",
       "      <td>FACEBASE:1-4GNG</td>\n",
       "      <td>None</td>\n",
       "      <td>FACEBASE:1-4GT6</td>\n",
       "      <td>FACEBASE:1-4GYR</td>\n",
       "      <td>None</td>\n",
       "      <td>FACEBASE:1-4GJA</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>3/30/15AL1-5</td>\n",
       "      <td>2016-06-20</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>https://auth.globus.org/b506963e-d274-11e5-99f...</td>\n",
       "      <td>2018-03-12T18:12:27.599487-07:00</td>\n",
       "      <td>2018-11-27T16:53:36.644948-08:00</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2YAT</td>\n",
       "      <td>TJP</td>\n",
       "      <td>A8IF3FR</td>\n",
       "      <td>NCBITAXON:10090</td>\n",
       "      <td>FACEBASE:1-4GNG</td>\n",
       "      <td>None</td>\n",
       "      <td>FACEBASE:1-4GT6</td>\n",
       "      <td>FACEBASE:1-4GYR</td>\n",
       "      <td>None</td>\n",
       "      <td>FACEBASE:1-4GJA</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>3/30/15AL1-1</td>\n",
       "      <td>2016-06-20</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>https://auth.globus.org/b506963e-d274-11e5-99f...</td>\n",
       "      <td>2018-03-12T18:12:27.599487-07:00</td>\n",
       "      <td>2018-11-27T16:53:35.57911-08:00</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>3Q1P</td>\n",
       "      <td>VHE</td>\n",
       "      <td>W8IF3SM</td>\n",
       "      <td>NCBITAXON:10090</td>\n",
       "      <td>FACEBASE:1-4GNG</td>\n",
       "      <td>None</td>\n",
       "      <td>FACEBASE:1-4GV2</td>\n",
       "      <td>FACEBASE:1-4GYR</td>\n",
       "      <td>None</td>\n",
       "      <td>FACEBASE:1-4GJA</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>4/14/14WL1-6</td>\n",
       "      <td>2014-04-14</td>\n",
       "      <td>None</td>\n",
       "      <td>https://auth.globus.org/b506963e-d274-11e5-99f...</td>\n",
       "      <td>https://auth.globus.org/b506963e-d274-11e5-99f...</td>\n",
       "      <td>2018-03-16T12:03:38.228586-07:00</td>\n",
       "      <td>2018-11-27T16:53:35.473682-08:00</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>3Q1T</td>\n",
       "      <td>VHE</td>\n",
       "      <td>W8IF4SM</td>\n",
       "      <td>NCBITAXON:10090</td>\n",
       "      <td>FACEBASE:1-4GNG</td>\n",
       "      <td>None</td>\n",
       "      <td>FACEBASE:1-4GV2</td>\n",
       "      <td>FACEBASE:1-4GYR</td>\n",
       "      <td>None</td>\n",
       "      <td>FACEBASE:1-4GJA</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>4/15/14WL1-7</td>\n",
       "      <td>2014-04-15</td>\n",
       "      <td>None</td>\n",
       "      <td>https://auth.globus.org/b506963e-d274-11e5-99f...</td>\n",
       "      <td>https://auth.globus.org/b506963e-d274-11e5-99f...</td>\n",
       "      <td>2018-03-16T12:03:38.228586-07:00</td>\n",
       "      <td>2018-11-27T16:53:35.473682-08:00</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>3Q1Y</td>\n",
       "      <td>VHE</td>\n",
       "      <td>W8IF5SM</td>\n",
       "      <td>NCBITAXON:10090</td>\n",
       "      <td>FACEBASE:1-4GNG</td>\n",
       "      <td>None</td>\n",
       "      <td>FACEBASE:1-4GV2</td>\n",
       "      <td>FACEBASE:1-4GYR</td>\n",
       "      <td>None</td>\n",
       "      <td>FACEBASE:1-4GJA</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>4/28/14WL1-1</td>\n",
       "      <td>2014-04-28</td>\n",
       "      <td>None</td>\n",
       "      <td>https://auth.globus.org/b506963e-d274-11e5-99f...</td>\n",
       "      <td>https://auth.globus.org/b506963e-d274-11e5-99f...</td>\n",
       "      <td>2018-03-16T12:03:38.228586-07:00</td>\n",
       "      <td>2018-11-27T16:53:35.473682-08:00</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>3V06</td>\n",
       "      <td>3TYP</td>\n",
       "      <td>scWCE18_S198</td>\n",
       "      <td>NCBITAXON:10090</td>\n",
       "      <td>FACEBASE:1-4GNR</td>\n",
       "      <td>None</td>\n",
       "      <td>FACEBASE:1-4GV2</td>\n",
       "      <td>FACEBASE:1-4GYR</td>\n",
       "      <td>None</td>\n",
       "      <td>FACEBASE:1-4GJA</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>6/6/17WL1,L2,L3</td>\n",
       "      <td>2017-06-06</td>\n",
       "      <td>None</td>\n",
       "      <td>https://auth.globus.org/f8ae714f-6015-48da-971...</td>\n",
       "      <td>https://auth.globus.org/b506963e-d274-11e5-99f...</td>\n",
       "      <td>2018-05-02T11:27:13.657825-07:00</td>\n",
       "      <td>2018-11-27T16:53:36.644948-08:00</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>3V1P</td>\n",
       "      <td>3V0E</td>\n",
       "      <td>scWLE18_S199</td>\n",
       "      <td>NCBITAXON:10090</td>\n",
       "      <td>FACEBASE:1-4GNR</td>\n",
       "      <td>None</td>\n",
       "      <td>FACEBASE:1-4GV2</td>\n",
       "      <td>FACEBASE:1-4GYR</td>\n",
       "      <td>None</td>\n",
       "      <td>FACEBASE:1-4GJA</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>6/6/17WL1,L2,L3</td>\n",
       "      <td>2017-06-06</td>\n",
       "      <td>None</td>\n",
       "      <td>https://auth.globus.org/8ae274db-d033-47eb-bd3...</td>\n",
       "      <td>https://auth.globus.org/b506963e-d274-11e5-99f...</td>\n",
       "      <td>2018-05-02T11:50:12.225872-07:00</td>\n",
       "      <td>2018-11-27T16:53:35.681942-08:00</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>64 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       RID dataset local_identifier          species         specimen  gene  \\\n",
       "0   1-4TT8  1-4TSP     scWFE18_S197  NCBITAXON:10090  FACEBASE:1-4GNR  None   \n",
       "1     2XEJ     TJP          A8IF4SM  NCBITAXON:10090  FACEBASE:1-4GNG  None   \n",
       "2     2XEP     TJP          A8IF1SM  NCBITAXON:10090  FACEBASE:1-4GNG  None   \n",
       "3     2XW6     TJP          A8IF4FR  NCBITAXON:10090  FACEBASE:1-4GNG  None   \n",
       "4     2YAT     TJP          A8IF3FR  NCBITAXON:10090  FACEBASE:1-4GNG  None   \n",
       "..     ...     ...              ...              ...              ...   ...   \n",
       "59    3Q1P     VHE          W8IF3SM  NCBITAXON:10090  FACEBASE:1-4GNG  None   \n",
       "60    3Q1T     VHE          W8IF4SM  NCBITAXON:10090  FACEBASE:1-4GNG  None   \n",
       "61    3Q1Y     VHE          W8IF5SM  NCBITAXON:10090  FACEBASE:1-4GNG  None   \n",
       "62    3V06    3TYP     scWCE18_S198  NCBITAXON:10090  FACEBASE:1-4GNR  None   \n",
       "63    3V1P    3V0E     scWLE18_S199  NCBITAXON:10090  FACEBASE:1-4GNR  None   \n",
       "\n",
       "           genotype           strain mutation            stage  ... gender  \\\n",
       "0   FACEBASE:1-4GV2  FACEBASE:1-4GYR     None  FACEBASE:1-4GJA  ...   None   \n",
       "1   FACEBASE:1-4GT6  FACEBASE:1-4GYR     None  FACEBASE:1-4GJA  ...   None   \n",
       "2   FACEBASE:1-4GT6  FACEBASE:1-4GYR     None  FACEBASE:1-4GJA  ...   None   \n",
       "3   FACEBASE:1-4GT6  FACEBASE:1-4GYR     None  FACEBASE:1-4GJA  ...   None   \n",
       "4   FACEBASE:1-4GT6  FACEBASE:1-4GYR     None  FACEBASE:1-4GJA  ...   None   \n",
       "..              ...              ...      ...              ...  ...    ...   \n",
       "59  FACEBASE:1-4GV2  FACEBASE:1-4GYR     None  FACEBASE:1-4GJA  ...   None   \n",
       "60  FACEBASE:1-4GV2  FACEBASE:1-4GYR     None  FACEBASE:1-4GJA  ...   None   \n",
       "61  FACEBASE:1-4GV2  FACEBASE:1-4GYR     None  FACEBASE:1-4GJA  ...   None   \n",
       "62  FACEBASE:1-4GV2  FACEBASE:1-4GYR     None  FACEBASE:1-4GJA  ...   None   \n",
       "63  FACEBASE:1-4GV2  FACEBASE:1-4GYR     None  FACEBASE:1-4GJA  ...   None   \n",
       "\n",
       "             litter collection_date _keywords  \\\n",
       "0              None            None      None   \n",
       "1      3/30/15AL1-5      2016-06-20      None   \n",
       "2      9/29/14AL1-5      2016-06-20      None   \n",
       "3      3/30/15AL1-5      2016-06-20      None   \n",
       "4      3/30/15AL1-1      2016-06-20      None   \n",
       "..              ...             ...       ...   \n",
       "59     4/14/14WL1-6      2014-04-14      None   \n",
       "60     4/15/14WL1-7      2014-04-15      None   \n",
       "61     4/28/14WL1-1      2014-04-28      None   \n",
       "62  6/6/17WL1,L2,L3      2017-06-06      None   \n",
       "63  6/6/17WL1,L2,L3      2017-06-06      None   \n",
       "\n",
       "                                                  RCB  \\\n",
       "0   https://auth.globus.org/8ae274db-d033-47eb-bd3...   \n",
       "1                                                None   \n",
       "2                                                None   \n",
       "3                                                None   \n",
       "4                                                None   \n",
       "..                                                ...   \n",
       "59  https://auth.globus.org/b506963e-d274-11e5-99f...   \n",
       "60  https://auth.globus.org/b506963e-d274-11e5-99f...   \n",
       "61  https://auth.globus.org/b506963e-d274-11e5-99f...   \n",
       "62  https://auth.globus.org/f8ae714f-6015-48da-971...   \n",
       "63  https://auth.globus.org/8ae274db-d033-47eb-bd3...   \n",
       "\n",
       "                                                  RMB  \\\n",
       "0   https://auth.globus.org/8ae274db-d033-47eb-bd3...   \n",
       "1   https://auth.globus.org/b506963e-d274-11e5-99f...   \n",
       "2   https://auth.globus.org/b506963e-d274-11e5-99f...   \n",
       "3   https://auth.globus.org/b506963e-d274-11e5-99f...   \n",
       "4   https://auth.globus.org/b506963e-d274-11e5-99f...   \n",
       "..                                                ...   \n",
       "59  https://auth.globus.org/b506963e-d274-11e5-99f...   \n",
       "60  https://auth.globus.org/b506963e-d274-11e5-99f...   \n",
       "61  https://auth.globus.org/b506963e-d274-11e5-99f...   \n",
       "62  https://auth.globus.org/b506963e-d274-11e5-99f...   \n",
       "63  https://auth.globus.org/b506963e-d274-11e5-99f...   \n",
       "\n",
       "                                 RCT                               RMT  \\\n",
       "0   2018-12-10T19:42:13.870864-08:00  2018-12-10T19:43:19.779894-08:00   \n",
       "1   2018-03-12T18:12:27.599487-07:00  2018-11-27T16:53:36.379287-08:00   \n",
       "2   2018-03-12T18:12:27.599487-07:00  2018-11-27T16:53:36.068132-08:00   \n",
       "3   2018-03-12T18:12:27.599487-07:00  2018-11-27T16:53:36.644948-08:00   \n",
       "4   2018-03-12T18:12:27.599487-07:00   2018-11-27T16:53:35.57911-08:00   \n",
       "..                               ...                               ...   \n",
       "59  2018-03-16T12:03:38.228586-07:00  2018-11-27T16:53:35.473682-08:00   \n",
       "60  2018-03-16T12:03:38.228586-07:00  2018-11-27T16:53:35.473682-08:00   \n",
       "61  2018-03-16T12:03:38.228586-07:00  2018-11-27T16:53:35.473682-08:00   \n",
       "62  2018-05-02T11:27:13.657825-07:00  2018-11-27T16:53:36.644948-08:00   \n",
       "63  2018-05-02T11:50:12.225872-07:00  2018-11-27T16:53:35.681942-08:00   \n",
       "\n",
       "   cell_source cell_characterization  \n",
       "0         None                  None  \n",
       "1         None                  None  \n",
       "2         None                  None  \n",
       "3         None                  None  \n",
       "4         None                  None  \n",
       "..         ...                   ...  \n",
       "59        None                  None  \n",
       "60        None                  None  \n",
       "61        None                  None  \n",
       "62        None                  None  \n",
       "63        None                  None  \n",
       "\n",
       "[64 rows x 24 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DataFrame(path.entities())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Example: filtering at different stages of the path\n",
    "Filtering a path does not have to be done at the end of a path. In fact, the initial intention of the ERMrest URI was to mimick \"RESTful\" semantics where a RESTful \"resource\" is identified, then filtered, then a \"sub-resource\" is identified, and then filtered, and so on."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://www.facebase.org/ermrest/catalog/1/entity/dataset:=isa:dataset/release_date::geq::2017-01-01/experiment:=isa:experiment/experiment_type=OBI%3A0001271/replicate:=isa:replicate/bioreplicate_number=1\n"
     ]
    }
   ],
   "source": [
    "path = dataset.filter(dataset.release_date >= '2017-01-01') \\\n",
    "    .link(experiment).filter(experiment.experiment_type == 'OBI:0001271') \\\n",
    "    .link(replicate).filter(replicate.bioreplicate_number == 1)\n",
    "    \n",
    "print(path.uri)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RID</th>\n",
       "      <th>dataset</th>\n",
       "      <th>biosample</th>\n",
       "      <th>bioreplicate_number</th>\n",
       "      <th>technical_replicate_number</th>\n",
       "      <th>RCB</th>\n",
       "      <th>RMB</th>\n",
       "      <th>RCT</th>\n",
       "      <th>RMT</th>\n",
       "      <th>experiment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-3T5A</td>\n",
       "      <td>1-3SWE</td>\n",
       "      <td>1-3T0A</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>https://auth.globus.org/f8ae714f-6015-48da-971...</td>\n",
       "      <td>https://auth.globus.org/f8ae714f-6015-48da-971...</td>\n",
       "      <td>2018-06-13T16:41:16.162698-07:00</td>\n",
       "      <td>2018-06-13T16:41:16.162698-07:00</td>\n",
       "      <td>1-3SZA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1-3T5E</td>\n",
       "      <td>1-3SWE</td>\n",
       "      <td>1-3T0E</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>https://auth.globus.org/f8ae714f-6015-48da-971...</td>\n",
       "      <td>https://auth.globus.org/f8ae714f-6015-48da-971...</td>\n",
       "      <td>2018-06-13T16:41:16.162698-07:00</td>\n",
       "      <td>2018-06-13T16:41:16.162698-07:00</td>\n",
       "      <td>1-3SZA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1-3T5J</td>\n",
       "      <td>1-3SWE</td>\n",
       "      <td>1-3T0J</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>https://auth.globus.org/f8ae714f-6015-48da-971...</td>\n",
       "      <td>https://auth.globus.org/f8ae714f-6015-48da-971...</td>\n",
       "      <td>2018-06-13T16:41:16.162698-07:00</td>\n",
       "      <td>2018-06-13T16:41:16.162698-07:00</td>\n",
       "      <td>1-3SZA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1-3T5P</td>\n",
       "      <td>1-3SWE</td>\n",
       "      <td>1-3T0P</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>https://auth.globus.org/f8ae714f-6015-48da-971...</td>\n",
       "      <td>https://auth.globus.org/f8ae714f-6015-48da-971...</td>\n",
       "      <td>2018-06-13T16:41:16.162698-07:00</td>\n",
       "      <td>2018-06-13T16:41:16.162698-07:00</td>\n",
       "      <td>1-3SZA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1-3T5T</td>\n",
       "      <td>1-3SWE</td>\n",
       "      <td>1-3T0T</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>https://auth.globus.org/f8ae714f-6015-48da-971...</td>\n",
       "      <td>https://auth.globus.org/f8ae714f-6015-48da-971...</td>\n",
       "      <td>2018-06-13T16:41:16.162698-07:00</td>\n",
       "      <td>2018-06-13T16:41:16.162698-07:00</td>\n",
       "      <td>1-3SZA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>274</th>\n",
       "      <td>3S5T</td>\n",
       "      <td>VQ4</td>\n",
       "      <td>3RZA</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>https://auth.globus.org/b506963e-d274-11e5-99f...</td>\n",
       "      <td>None</td>\n",
       "      <td>2018-03-26T16:34:43.289759-07:00</td>\n",
       "      <td>2018-06-05T18:44:38.502116-07:00</td>\n",
       "      <td>3S1T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>275</th>\n",
       "      <td>3S6A</td>\n",
       "      <td>VQ4</td>\n",
       "      <td>3RZT</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>https://auth.globus.org/b506963e-d274-11e5-99f...</td>\n",
       "      <td>None</td>\n",
       "      <td>2018-03-26T16:34:43.289759-07:00</td>\n",
       "      <td>2018-06-05T18:44:38.502116-07:00</td>\n",
       "      <td>3S1T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>276</th>\n",
       "      <td>3S6E</td>\n",
       "      <td>VQ4</td>\n",
       "      <td>3RZY</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>https://auth.globus.org/b506963e-d274-11e5-99f...</td>\n",
       "      <td>None</td>\n",
       "      <td>2018-03-26T16:36:10.863089-07:00</td>\n",
       "      <td>2018-06-05T18:44:38.502116-07:00</td>\n",
       "      <td>3S1Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>277</th>\n",
       "      <td>3TSE</td>\n",
       "      <td>2ARP</td>\n",
       "      <td>30AA</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>https://auth.globus.org/b506963e-d274-11e5-99f...</td>\n",
       "      <td>None</td>\n",
       "      <td>2018-04-11T12:26:16.576316-07:00</td>\n",
       "      <td>2018-06-05T18:44:38.502116-07:00</td>\n",
       "      <td>341Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>278</th>\n",
       "      <td>3TSJ</td>\n",
       "      <td>2ARP</td>\n",
       "      <td>30GP</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>https://auth.globus.org/b506963e-d274-11e5-99f...</td>\n",
       "      <td>None</td>\n",
       "      <td>2018-04-11T12:28:24.259798-07:00</td>\n",
       "      <td>2018-06-05T18:44:38.502116-07:00</td>\n",
       "      <td>3422</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>279 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        RID dataset biosample  bioreplicate_number  \\\n",
       "0    1-3T5A  1-3SWE    1-3T0A                    1   \n",
       "1    1-3T5E  1-3SWE    1-3T0E                    1   \n",
       "2    1-3T5J  1-3SWE    1-3T0J                    1   \n",
       "3    1-3T5P  1-3SWE    1-3T0P                    1   \n",
       "4    1-3T5T  1-3SWE    1-3T0T                    1   \n",
       "..      ...     ...       ...                  ...   \n",
       "274    3S5T     VQ4      3RZA                    1   \n",
       "275    3S6A     VQ4      3RZT                    1   \n",
       "276    3S6E     VQ4      3RZY                    1   \n",
       "277    3TSE    2ARP      30AA                    1   \n",
       "278    3TSJ    2ARP      30GP                    1   \n",
       "\n",
       "     technical_replicate_number  \\\n",
       "0                             1   \n",
       "1                             1   \n",
       "2                             1   \n",
       "3                             1   \n",
       "4                             1   \n",
       "..                          ...   \n",
       "274                           1   \n",
       "275                           1   \n",
       "276                           1   \n",
       "277                           1   \n",
       "278                           1   \n",
       "\n",
       "                                                   RCB  \\\n",
       "0    https://auth.globus.org/f8ae714f-6015-48da-971...   \n",
       "1    https://auth.globus.org/f8ae714f-6015-48da-971...   \n",
       "2    https://auth.globus.org/f8ae714f-6015-48da-971...   \n",
       "3    https://auth.globus.org/f8ae714f-6015-48da-971...   \n",
       "4    https://auth.globus.org/f8ae714f-6015-48da-971...   \n",
       "..                                                 ...   \n",
       "274  https://auth.globus.org/b506963e-d274-11e5-99f...   \n",
       "275  https://auth.globus.org/b506963e-d274-11e5-99f...   \n",
       "276  https://auth.globus.org/b506963e-d274-11e5-99f...   \n",
       "277  https://auth.globus.org/b506963e-d274-11e5-99f...   \n",
       "278  https://auth.globus.org/b506963e-d274-11e5-99f...   \n",
       "\n",
       "                                                   RMB  \\\n",
       "0    https://auth.globus.org/f8ae714f-6015-48da-971...   \n",
       "1    https://auth.globus.org/f8ae714f-6015-48da-971...   \n",
       "2    https://auth.globus.org/f8ae714f-6015-48da-971...   \n",
       "3    https://auth.globus.org/f8ae714f-6015-48da-971...   \n",
       "4    https://auth.globus.org/f8ae714f-6015-48da-971...   \n",
       "..                                                 ...   \n",
       "274                                               None   \n",
       "275                                               None   \n",
       "276                                               None   \n",
       "277                                               None   \n",
       "278                                               None   \n",
       "\n",
       "                                  RCT                               RMT  \\\n",
       "0    2018-06-13T16:41:16.162698-07:00  2018-06-13T16:41:16.162698-07:00   \n",
       "1    2018-06-13T16:41:16.162698-07:00  2018-06-13T16:41:16.162698-07:00   \n",
       "2    2018-06-13T16:41:16.162698-07:00  2018-06-13T16:41:16.162698-07:00   \n",
       "3    2018-06-13T16:41:16.162698-07:00  2018-06-13T16:41:16.162698-07:00   \n",
       "4    2018-06-13T16:41:16.162698-07:00  2018-06-13T16:41:16.162698-07:00   \n",
       "..                                ...                               ...   \n",
       "274  2018-03-26T16:34:43.289759-07:00  2018-06-05T18:44:38.502116-07:00   \n",
       "275  2018-03-26T16:34:43.289759-07:00  2018-06-05T18:44:38.502116-07:00   \n",
       "276  2018-03-26T16:36:10.863089-07:00  2018-06-05T18:44:38.502116-07:00   \n",
       "277  2018-04-11T12:26:16.576316-07:00  2018-06-05T18:44:38.502116-07:00   \n",
       "278  2018-04-11T12:28:24.259798-07:00  2018-06-05T18:44:38.502116-07:00   \n",
       "\n",
       "    experiment  \n",
       "0       1-3SZA  \n",
       "1       1-3SZA  \n",
       "2       1-3SZA  \n",
       "3       1-3SZA  \n",
       "4       1-3SZA  \n",
       "..         ...  \n",
       "274       3S1T  \n",
       "275       3S1T  \n",
       "276       3S1Y  \n",
       "277       341Y  \n",
       "278       3422  \n",
       "\n",
       "[279 rows x 10 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DataFrame(path.entities())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Linking Examples\n",
    "\n",
    "### Example: explicit column links\n",
    "Up until now, the examples have shown how to link entities via _implicit_ join predicates. That is, we knew there existed a foriegn key reference constraint between foreign keys of one entity and keys of another entity. We needed only to ask ERMrest to link the entities in order to get the linked set.\n",
    "\n",
    "The problem with implicit links is that it become _ambiguous_ if there are more than one foreign key reference between tables. To support these situations, ERMrest and the `DataPath`'s `link(...)` method can specify the columns to use for the link condition, explicitly.\n",
    "\n",
    "The structure of the `on` clause is:\n",
    "- an equality comparison operation where\n",
    "- the _left_ operand is a column of the _left_ table instance which is also the path _context_ before the link method is called, and\n",
    "- the _right_ operand is a column of the _right_ table instance which is the table _to be linked_ to the path."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://www.facebase.org/ermrest/catalog/1/entity/dataset:=isa:dataset/experiment:=(RID)=(isa:experiment:dataset)\n"
     ]
    }
   ],
   "source": [
    "path = dataset.link(experiment, on=(dataset.RID==experiment.dataset))\n",
    "print(path.uri)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**IMPORTANT** Not all tables are related by foreign key references. ERMrest does not allow arbitrary relational joins. Tables must be related by a foreign key reference in order to link them in a data path."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RID</th>\n",
       "      <th>dataset</th>\n",
       "      <th>local_identifier</th>\n",
       "      <th>experiment_type</th>\n",
       "      <th>molecule_type</th>\n",
       "      <th>strandedness</th>\n",
       "      <th>rnaseq_selection</th>\n",
       "      <th>target_of_assay</th>\n",
       "      <th>chromatin_modifier</th>\n",
       "      <th>transcription_factor</th>\n",
       "      <th>histone_modification</th>\n",
       "      <th>control_assay</th>\n",
       "      <th>protocol</th>\n",
       "      <th>RCB</th>\n",
       "      <th>RMB</th>\n",
       "      <th>RCT</th>\n",
       "      <th>RMT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-3SD2</td>\n",
       "      <td>1-3SB2</td>\n",
       "      <td>3XhET4_pHsp68-lacZ-tdTomato</td>\n",
       "      <td>OBI:0002083</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>https://auth.globus.org/a1d30d14-b3b0-49de-854...</td>\n",
       "      <td>https://auth.globus.org/b506963e-d274-11e5-99f...</td>\n",
       "      <td>2018-06-08T12:40:48.98831-07:00</td>\n",
       "      <td>2018-11-27T16:54:17.199655-08:00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1-3SHT</td>\n",
       "      <td>1-3SGA</td>\n",
       "      <td>3XcET4_pHsp68-lacZ-tdTomato</td>\n",
       "      <td>OBI:0002083</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>https://auth.globus.org/a1d30d14-b3b0-49de-854...</td>\n",
       "      <td>https://auth.globus.org/b506963e-d274-11e5-99f...</td>\n",
       "      <td>2018-06-08T14:29:49.540747-07:00</td>\n",
       "      <td>2018-11-27T16:54:17.199655-08:00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1-3SNY</td>\n",
       "      <td>1-3SM2</td>\n",
       "      <td>3XhET7_pHsp68-lacZ-tdTomato</td>\n",
       "      <td>OBI:0002083</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>https://auth.globus.org/a1d30d14-b3b0-49de-854...</td>\n",
       "      <td>https://auth.globus.org/b506963e-d274-11e5-99f...</td>\n",
       "      <td>2018-06-08T14:47:53.103435-07:00</td>\n",
       "      <td>2018-11-27T16:54:17.199655-08:00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      RID dataset             local_identifier experiment_type molecule_type  \\\n",
       "0  1-3SD2  1-3SB2  3XhET4_pHsp68-lacZ-tdTomato     OBI:0002083          None   \n",
       "1  1-3SHT  1-3SGA  3XcET4_pHsp68-lacZ-tdTomato     OBI:0002083          None   \n",
       "2  1-3SNY  1-3SM2  3XhET7_pHsp68-lacZ-tdTomato     OBI:0002083          None   \n",
       "\n",
       "  strandedness rnaseq_selection target_of_assay chromatin_modifier  \\\n",
       "0         None             None            None               None   \n",
       "1         None             None            None               None   \n",
       "2         None             None            None               None   \n",
       "\n",
       "  transcription_factor histone_modification control_assay protocol  \\\n",
       "0                 None                 None          None     None   \n",
       "1                 None                 None          None     None   \n",
       "2                 None                 None          None     None   \n",
       "\n",
       "                                                 RCB  \\\n",
       "0  https://auth.globus.org/a1d30d14-b3b0-49de-854...   \n",
       "1  https://auth.globus.org/a1d30d14-b3b0-49de-854...   \n",
       "2  https://auth.globus.org/a1d30d14-b3b0-49de-854...   \n",
       "\n",
       "                                                 RMB  \\\n",
       "0  https://auth.globus.org/b506963e-d274-11e5-99f...   \n",
       "1  https://auth.globus.org/b506963e-d274-11e5-99f...   \n",
       "2  https://auth.globus.org/b506963e-d274-11e5-99f...   \n",
       "\n",
       "                                RCT                               RMT  \n",
       "0   2018-06-08T12:40:48.98831-07:00  2018-11-27T16:54:17.199655-08:00  \n",
       "1  2018-06-08T14:29:49.540747-07:00  2018-11-27T16:54:17.199655-08:00  \n",
       "2  2018-06-08T14:47:53.103435-07:00  2018-11-27T16:54:17.199655-08:00  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DataFrame(path.entities().fetch(limit=3))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Example: explicit column links combined with table aliasing\n",
    "As usual, table instances are generated automatically unless we provide a table alias."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://www.facebase.org/ermrest/catalog/1/entity/dataset:=isa:dataset/S:=(RID)=(isa:biosample:dataset)\n"
     ]
    }
   ],
   "source": [
    "path = dataset.link(biosample.alias('S'), on=(dataset.RID==biosample.dataset))\n",
    "print(path.uri)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notice that we cannot use the alias right away in the `on` clause because it was not _bound_ to the path until _after_ the `link(...)` operation was performed."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Example: links with \"outer join\" semantics\n",
    "Up until now, the examples have shown \"`link`s\" with _inner join_ semantics. _Outer join_ semantics can be expressed as part of explicit column links, and _only_ when using explicit column links.\n",
    "\n",
    "The `link(...)` method accepts a \"`join_type`\" parameter, i.e., \"`.link(... join_type=TYPE)`\", where _TYPE_ may be `'left'`, `'right'`, `'full'`, and defaults to `''` which indicates inner join type.\n",
    "\n",
    "By '`left`' outer joining in the link from `'dataset'` to `'experiment`' and to `'biosample'`, and then reseting the context of the path to `'dataset'`, the following path gives us a reference to `'dataset'` entities that _whether or not_ they have any experiments or biosamples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://www.facebase.org/ermrest/catalog/1/attribute/dataset:=isa:dataset/E:=left(RID)=(isa:experiment:dataset)/$dataset/S:=left(RID)=(isa:biosample:dataset)/$dataset/dataset:RID,dataset:title,E:experiment_type,S:species\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "855"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Notice in between `link`s that we have to reset the context back to `dataset` so that the\n",
    "# second join is also left joined from the dataset table instance.\n",
    "path = dataset.link(experiment.alias('E'), on=dataset.RID==experiment.dataset, join_type='left') \\\n",
    "              .dataset \\\n",
    "              .link(biosample.alias('S'), on=dataset.RID==biosample.dataset, join_type='left') \\\n",
    "\n",
    "# Notice that we have to perform the attribute fetch from the context of the `path.dataset`\n",
    "# table instance.\n",
    "results = path.dataset.attributes(path.dataset.RID, \n",
    "                                  path.dataset.title, \n",
    "                                  path.E.experiment_type, \n",
    "                                  path.S.species)\n",
    "\n",
    "print(results.uri)\n",
    "len(results)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can see above that we have a full set of datasets _whether or not_ they have any experiments with biosamples. For further evidence, we can convert to a DataFrame and look at a slice of its entries. Note that the biosample's 'species' and 'stage' attributes do not exist for some results (i.e., `NaN`) because those attributes did not exist for the join condition."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RID</th>\n",
       "      <th>title</th>\n",
       "      <th>experiment_type</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TZP</td>\n",
       "      <td>Gene expression microarray - mouse E10.5 maxil...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>VTG</td>\n",
       "      <td>Gene expression microarray - mouse E10.5 mandi...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>V4G</td>\n",
       "      <td>miRNAs expressed in the E13.5 mouse mandible</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TMP</td>\n",
       "      <td>miRNAs expressed in the 72hpf zebrafish head</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TVA</td>\n",
       "      <td>Gene expression microarray - mouse E10.5 media...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>V1C</td>\n",
       "      <td>Gene expression microarray - mouse E11.5 singl...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>V8A</td>\n",
       "      <td>Gene expression microarray - mouse E10.5 later...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>VM6</td>\n",
       "      <td>Gene expression microarray - mouse E10.5 media...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>VHR</td>\n",
       "      <td>Gene expression profiling of palatal and mandi...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>V9A</td>\n",
       "      <td>Gene expression microarray - mouse E10.5 olfac...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   RID                                              title experiment_type  \\\n",
       "0  TZP  Gene expression microarray - mouse E10.5 maxil...            None   \n",
       "1  VTG  Gene expression microarray - mouse E10.5 mandi...            None   \n",
       "2  V4G       miRNAs expressed in the E13.5 mouse mandible            None   \n",
       "3  TMP       miRNAs expressed in the 72hpf zebrafish head            None   \n",
       "4  TVA  Gene expression microarray - mouse E10.5 media...            None   \n",
       "5  V1C  Gene expression microarray - mouse E11.5 singl...            None   \n",
       "6  V8A  Gene expression microarray - mouse E10.5 later...            None   \n",
       "7  VM6  Gene expression microarray - mouse E10.5 media...            None   \n",
       "8  VHR  Gene expression profiling of palatal and mandi...            None   \n",
       "9  V9A  Gene expression microarray - mouse E10.5 olfac...            None   \n",
       "\n",
       "  species  \n",
       "0    None  \n",
       "1    None  \n",
       "2    None  \n",
       "3    None  \n",
       "4    None  \n",
       "5    None  \n",
       "6    None  \n",
       "7    None  \n",
       "8    None  \n",
       "9    None  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DataFrame(results)[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Faceting Examples\n",
    "You may have noticed that in the examples above, the 'species' and 'experiment_type' attributes are identifiers ('CURIE's to be precise). We may want to construct filters on our datasets based on these categories. This can be used for \"faceted search\" modes and can be useful even within the context of programmatic access to data in the catalog."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Example: faceting on \"related\" tables\n",
    "Let's say we want to find all of the biosamples in our catalog where their species are 'Mus musculus' and their age stage are 'E10.5'.\n",
    "\n",
    "We need to extend our understanding of the data model with the following tables that are related to '`biosample`'.\n",
    "- `isa.biosample.species -> vocab.species`: the biosample table has a foreign key reference to the '`species`' table.\n",
    "- `isa.biosample.stage -> vocab.stage`: the biosample table has a foreign key reference to the '`stage`' table.\n",
    "\n",
    "We may say that `species` and `stage` are _related_ to the `biosample` table in the sense that `biosample` has a direct foreign key relationship from it to them.\n",
    "\n",
    "For convenience, we will get local variables for the species and stage tables."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "species = pb.vocab.species\n",
    "stage = pb.vocab.stage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, let's link samples with species and filter on the term \"Mus musculus\" (i.e., \"mouse\")."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://www.facebase.org/ermrest/catalog/1/entity/S:=isa:biosample/species:=vocab:species/name=Mus%20musculus\n"
     ]
    }
   ],
   "source": [
    "# Here we have to use the container `columns_definitions` because `name` is reserved\n",
    "path = biosample.alias('S').link(species).filter(species.column_definitions['name'] == 'Mus musculus')\n",
    "print(path.uri)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now the _context_ of the path is the `species` table instance, but we need to link from the `biosample` to the age `stage` table.\n",
    "\n",
    "To do so, we reference the `biosample` table instance, in this case using its alias `S`. Then we link off of that table instance which updates the `path` itself."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://www.facebase.org/ermrest/catalog/1/entity/S:=isa:biosample/species:=vocab:species/name=Mus%20musculus/$S/stage:=vocab:stage/name=E10.5\n"
     ]
    }
   ],
   "source": [
    "path.S.link(stage).filter(stage.column_definitions['name'] == 'E10.5')\n",
    "print(path.uri)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, the path _context_ is the age `stage` table instance, but we wanted to get the entities for the `biosample` table. To do so, again we will reference the `biosample` table instance by the alias `S` we used. From there, we will call the `entities(...)` method to get the samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://www.facebase.org/ermrest/catalog/1/attribute/S:=isa:biosample/species:=vocab:species/name=Mus%20musculus/$S/stage:=vocab:stage/name=E10.5/$S/S:RID,S:collection_date,species:=species:name,species_uri:=species:uri,stage:=stage:name,stage_uri:=stage:uri\n"
     ]
    }
   ],
   "source": [
    "results = path.S.attributes(path.S.RID,\n",
    "                            path.S.collection_date,\n",
    "                            path.species.column_definitions['name'].alias('species'),\n",
    "                            path.species.column_definitions['uri'].alias('species_uri'),\n",
    "                            path.stage.column_definitions['name'].alias('stage'),\n",
    "                            path.stage.column_definitions['uri'].alias('stage_uri'))\n",
    "print(results.uri)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RID</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>species</th>\n",
       "      <th>species_uri</th>\n",
       "      <th>stage</th>\n",
       "      <th>stage_uri</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-4TQ8</td>\n",
       "      <td>None</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>https://www.facebase.org/id/1-4FZJ</td>\n",
       "      <td>E10.5</td>\n",
       "      <td>https://www.facebase.org/id/1-4GJ0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1-4TR6</td>\n",
       "      <td>None</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>https://www.facebase.org/id/1-4FZJ</td>\n",
       "      <td>E10.5</td>\n",
       "      <td>https://www.facebase.org/id/1-4GJ0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1-4TRG</td>\n",
       "      <td>None</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>https://www.facebase.org/id/1-4FZJ</td>\n",
       "      <td>E10.5</td>\n",
       "      <td>https://www.facebase.org/id/1-4GJ0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2XHT</td>\n",
       "      <td>2015-07-14</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>https://www.facebase.org/id/1-4FZJ</td>\n",
       "      <td>E10.5</td>\n",
       "      <td>https://www.facebase.org/id/1-4GJ0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2XST</td>\n",
       "      <td>2015-07-14</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>https://www.facebase.org/id/1-4FZJ</td>\n",
       "      <td>E10.5</td>\n",
       "      <td>https://www.facebase.org/id/1-4GJ0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>32HP</td>\n",
       "      <td>2015-07-14</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>https://www.facebase.org/id/1-4FZJ</td>\n",
       "      <td>E10.5</td>\n",
       "      <td>https://www.facebase.org/id/1-4GJ0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>3RPT</td>\n",
       "      <td>None</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>https://www.facebase.org/id/1-4FZJ</td>\n",
       "      <td>E10.5</td>\n",
       "      <td>https://www.facebase.org/id/1-4GJ0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>3T3P</td>\n",
       "      <td>None</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>https://www.facebase.org/id/1-4FZJ</td>\n",
       "      <td>E10.5</td>\n",
       "      <td>https://www.facebase.org/id/1-4GJ0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>3T3T</td>\n",
       "      <td>None</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>https://www.facebase.org/id/1-4FZJ</td>\n",
       "      <td>E10.5</td>\n",
       "      <td>https://www.facebase.org/id/1-4GJ0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>3T3Y</td>\n",
       "      <td>None</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>https://www.facebase.org/id/1-4FZJ</td>\n",
       "      <td>E10.5</td>\n",
       "      <td>https://www.facebase.org/id/1-4GJ0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>61 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       RID collection_date       species                         species_uri  \\\n",
       "0   1-4TQ8            None  Mus musculus  https://www.facebase.org/id/1-4FZJ   \n",
       "1   1-4TR6            None  Mus musculus  https://www.facebase.org/id/1-4FZJ   \n",
       "2   1-4TRG            None  Mus musculus  https://www.facebase.org/id/1-4FZJ   \n",
       "3     2XHT      2015-07-14  Mus musculus  https://www.facebase.org/id/1-4FZJ   \n",
       "4     2XST      2015-07-14  Mus musculus  https://www.facebase.org/id/1-4FZJ   \n",
       "..     ...             ...           ...                                 ...   \n",
       "56    32HP      2015-07-14  Mus musculus  https://www.facebase.org/id/1-4FZJ   \n",
       "57    3RPT            None  Mus musculus  https://www.facebase.org/id/1-4FZJ   \n",
       "58    3T3P            None  Mus musculus  https://www.facebase.org/id/1-4FZJ   \n",
       "59    3T3T            None  Mus musculus  https://www.facebase.org/id/1-4FZJ   \n",
       "60    3T3Y            None  Mus musculus  https://www.facebase.org/id/1-4FZJ   \n",
       "\n",
       "    stage                           stage_uri  \n",
       "0   E10.5  https://www.facebase.org/id/1-4GJ0  \n",
       "1   E10.5  https://www.facebase.org/id/1-4GJ0  \n",
       "2   E10.5  https://www.facebase.org/id/1-4GJ0  \n",
       "3   E10.5  https://www.facebase.org/id/1-4GJ0  \n",
       "4   E10.5  https://www.facebase.org/id/1-4GJ0  \n",
       "..    ...                                 ...  \n",
       "56  E10.5  https://www.facebase.org/id/1-4GJ0  \n",
       "57  E10.5  https://www.facebase.org/id/1-4GJ0  \n",
       "58  E10.5  https://www.facebase.org/id/1-4GJ0  \n",
       "59  E10.5  https://www.facebase.org/id/1-4GJ0  \n",
       "60  E10.5  https://www.facebase.org/id/1-4GJ0  \n",
       "\n",
       "[61 rows x 6 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DataFrame(results)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Grouping Examples\n",
    "Now support you would like to aggregate all of the vocabulary terms associated with a Dataset. Here, we examine what happens when you have a model such that `dataset <- dataset_VOCAB -> VOCAB` where `VOCAB` is a placeholder for a table that includes a vocabulary term set. These tables typically have a `name` column for the human-readable preferred label to go along with the formal URI or CURIE of the concept class."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://www.facebase.org/ermrest/catalog/1/attributegroup/dataset:=isa:dataset/dataset_organism:=left(id)=(isa:dataset_organism:dataset_id)/species:=vocab:species/$dataset/dataset_experiment_type:=left(id)=(isa:dataset_experiment_type:dataset_id)/experiment_type:=vocab:experiment_type/$dataset/RID;title,species:=array_d(species:name),experiment_type:=array_d(experiment_type:name)\n",
      "855\n"
     ]
    }
   ],
   "source": [
    "# We need to import the `ArrayD` aggregate function for this example.\n",
    "from deriva.core.datapath import ArrayD\n",
    "\n",
    "# For convenience, get python objects for the additional tables.\n",
    "dataset_organism = pb.isa.dataset_organism\n",
    "dataset_experiment_type = pb.isa.dataset_experiment_type\n",
    "species = pb.vocab.species\n",
    "experiment_type = pb.vocab.experiment_type\n",
    "\n",
    "# Start by doing a couple left outer joins on the dataset-term association tables, then link\n",
    "# (i.e., inner join) the associated vocabulary term table, then reset the context back to the\n",
    "# dataset table.\n",
    "path = dataset.link(dataset_organism, on=dataset.id==dataset_organism.dataset_id, join_type='left') \\\n",
    "              .link(species) \\\n",
    "              .dataset \\\n",
    "              .link(dataset_experiment_type, on=dataset.id==dataset_experiment_type.dataset_id, join_type='left') \\\n",
    "              .link(experiment_type)\n",
    "\n",
    "# Again, notice that we reset the context to the `dataset` table alias so that we will retrieve \n",
    "# dataset entities based on the groupings to be defined next. For the groupby key we will use the\n",
    "# dataset.RID, but for this example any primary key would work. Then we will get aggregate arrays\n",
    "# of the linked vocabulary tables.\n",
    "results = path.dataset.groupby(dataset.RID).attributes(\n",
    "    dataset.title,\n",
    "    ArrayD(path.species.column_definitions['name']).alias('species'),\n",
    "    ArrayD(path.experiment_type.column_definitions['name']).alias('experiment_type')\n",
    ")\n",
    "\n",
    "#results = path.dataset.entities()\n",
    "print(results.uri)\n",
    "print(len(results))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RID</th>\n",
       "      <th>species</th>\n",
       "      <th>experiment_type</th>\n",
       "      <th>title</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-3SB2</td>\n",
       "      <td>[Mus musculus]</td>\n",
       "      <td>[enhancer activity detection by reporter gene ...</td>\n",
       "      <td>Activity of human neural crest enhancer near G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1-3SGA</td>\n",
       "      <td>[Mus musculus]</td>\n",
       "      <td>[enhancer activity detection by reporter gene ...</td>\n",
       "      <td>Activity of chimp neural crest enhancer near G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1-3SM2</td>\n",
       "      <td>[Mus musculus]</td>\n",
       "      <td>[enhancer activity detection by reporter gene ...</td>\n",
       "      <td>Activity of human neural crest enhancer near F...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1-3SQJ</td>\n",
       "      <td>[Mus musculus]</td>\n",
       "      <td>[enhancer activity detection by reporter gene ...</td>\n",
       "      <td>Activity of chimp neural crest enhancer near F...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1-3SVP</td>\n",
       "      <td>[Homo sapiens]</td>\n",
       "      <td>[exome sequencing assay, genotyping assay, ima...</td>\n",
       "      <td>FB0023_18mo male with hypertelorism_Candidate ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1-3SVT</td>\n",
       "      <td>[Homo sapiens]</td>\n",
       "      <td>[exome sequencing assay, genotyping assay, ima...</td>\n",
       "      <td>FB0043_3 male cousins with natal teeth and ano...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1-3SVY</td>\n",
       "      <td>[Homo sapiens]</td>\n",
       "      <td>[exome sequencing assay, genotyping assay, ima...</td>\n",
       "      <td>FB0051_6 year old male with hypertelorism (mar...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1-3SW2</td>\n",
       "      <td>[Homo sapiens]</td>\n",
       "      <td>[exome sequencing assay, genotyping assay, ima...</td>\n",
       "      <td>FB0064_Male with Congenital craniosynostosis_C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1-3SW6</td>\n",
       "      <td>[Homo sapiens]</td>\n",
       "      <td>[exome sequencing assay, genotyping assay, ima...</td>\n",
       "      <td>FB0115_11yo male with dysmorphic facial featur...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1-3SWA</td>\n",
       "      <td>[Homo sapiens]</td>\n",
       "      <td>[exome sequencing assay, genotyping assay, ima...</td>\n",
       "      <td>FB0122_29yo female with bilateral hearing loss...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1-3SWE</td>\n",
       "      <td>[Mus musculus]</td>\n",
       "      <td>[RNA-seq assay]</td>\n",
       "      <td>Interpalatine suture, WT and Fgfr2+/S252W (Ape...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1-3SXY</td>\n",
       "      <td>[Mus musculus]</td>\n",
       "      <td>[RNA-seq assay]</td>\n",
       "      <td>Maxillary-Palatine Suture, WT and Fgfr2+/S252W...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1-3V7A</td>\n",
       "      <td>[Homo sapiens]</td>\n",
       "      <td>[exome sequencing assay, genotyping assay, ima...</td>\n",
       "      <td>FB0011_9 year old female with single central m...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1-3V7J</td>\n",
       "      <td>[Homo sapiens]</td>\n",
       "      <td>[exome sequencing assay, genotyping assay, ima...</td>\n",
       "      <td>FB0015_Male with bilateral cleft lip and palate</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1-3V7P</td>\n",
       "      <td>[Homo sapiens]</td>\n",
       "      <td>[exome sequencing assay, genotyping assay, ima...</td>\n",
       "      <td>FB0027_29 month old male_Multiple congenital a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1-3V8A</td>\n",
       "      <td>[Homo sapiens]</td>\n",
       "      <td>[exome sequencing assay, genotyping assay, ima...</td>\n",
       "      <td>FB0039_\"Identical Male Twins\"_Treacher Collins...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1-3V8Y</td>\n",
       "      <td>[Homo sapiens]</td>\n",
       "      <td>[genotyping assay, imaging assay, Morphometric...</td>\n",
       "      <td>FB0041_8 yo male_Brachi-Oto-Renal (BOR) Syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1-3V9E</td>\n",
       "      <td>[Homo sapiens]</td>\n",
       "      <td>[exome sequencing assay, genotyping assay, ima...</td>\n",
       "      <td>FB0049_17 month old male with bilateral macros...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1-3VAT</td>\n",
       "      <td>[Homo sapiens]</td>\n",
       "      <td>[exome sequencing assay, genotyping assay, ima...</td>\n",
       "      <td>FB0062_Proband Only Saliva_Blepharo-cheilo-don...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1-3VBE</td>\n",
       "      <td>[Homo sapiens]</td>\n",
       "      <td>[exome sequencing assay, genotyping assay, ima...</td>\n",
       "      <td>FB0124_Trio Saliva_5 mo male with Hallermann S...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       RID         species                                    experiment_type  \\\n",
       "0   1-3SB2  [Mus musculus]  [enhancer activity detection by reporter gene ...   \n",
       "1   1-3SGA  [Mus musculus]  [enhancer activity detection by reporter gene ...   \n",
       "2   1-3SM2  [Mus musculus]  [enhancer activity detection by reporter gene ...   \n",
       "3   1-3SQJ  [Mus musculus]  [enhancer activity detection by reporter gene ...   \n",
       "4   1-3SVP  [Homo sapiens]  [exome sequencing assay, genotyping assay, ima...   \n",
       "5   1-3SVT  [Homo sapiens]  [exome sequencing assay, genotyping assay, ima...   \n",
       "6   1-3SVY  [Homo sapiens]  [exome sequencing assay, genotyping assay, ima...   \n",
       "7   1-3SW2  [Homo sapiens]  [exome sequencing assay, genotyping assay, ima...   \n",
       "8   1-3SW6  [Homo sapiens]  [exome sequencing assay, genotyping assay, ima...   \n",
       "9   1-3SWA  [Homo sapiens]  [exome sequencing assay, genotyping assay, ima...   \n",
       "10  1-3SWE  [Mus musculus]                                    [RNA-seq assay]   \n",
       "11  1-3SXY  [Mus musculus]                                    [RNA-seq assay]   \n",
       "12  1-3V7A  [Homo sapiens]  [exome sequencing assay, genotyping assay, ima...   \n",
       "13  1-3V7J  [Homo sapiens]  [exome sequencing assay, genotyping assay, ima...   \n",
       "14  1-3V7P  [Homo sapiens]  [exome sequencing assay, genotyping assay, ima...   \n",
       "15  1-3V8A  [Homo sapiens]  [exome sequencing assay, genotyping assay, ima...   \n",
       "16  1-3V8Y  [Homo sapiens]  [genotyping assay, imaging assay, Morphometric...   \n",
       "17  1-3V9E  [Homo sapiens]  [exome sequencing assay, genotyping assay, ima...   \n",
       "18  1-3VAT  [Homo sapiens]  [exome sequencing assay, genotyping assay, ima...   \n",
       "19  1-3VBE  [Homo sapiens]  [exome sequencing assay, genotyping assay, ima...   \n",
       "\n",
       "                                                title  \n",
       "0   Activity of human neural crest enhancer near G...  \n",
       "1   Activity of chimp neural crest enhancer near G...  \n",
       "2   Activity of human neural crest enhancer near F...  \n",
       "3   Activity of chimp neural crest enhancer near F...  \n",
       "4   FB0023_18mo male with hypertelorism_Candidate ...  \n",
       "5   FB0043_3 male cousins with natal teeth and ano...  \n",
       "6   FB0051_6 year old male with hypertelorism (mar...  \n",
       "7   FB0064_Male with Congenital craniosynostosis_C...  \n",
       "8   FB0115_11yo male with dysmorphic facial featur...  \n",
       "9   FB0122_29yo female with bilateral hearing loss...  \n",
       "10  Interpalatine suture, WT and Fgfr2+/S252W (Ape...  \n",
       "11  Maxillary-Palatine Suture, WT and Fgfr2+/S252W...  \n",
       "12  FB0011_9 year old female with single central m...  \n",
       "13    FB0015_Male with bilateral cleft lip and palate  \n",
       "14  FB0027_29 month old male_Multiple congenital a...  \n",
       "15  FB0039_\"Identical Male Twins\"_Treacher Collins...  \n",
       "16   FB0041_8 yo male_Brachi-Oto-Renal (BOR) Syndrome  \n",
       "17  FB0049_17 month old male with bilateral macros...  \n",
       "18  FB0062_Proband Only Saliva_Blepharo-cheilo-don...  \n",
       "19  FB0124_Trio Saliva_5 mo male with Hallermann S...  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DataFrame(results.fetch(limit=20))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
